File: lab-6-6/src/main.scala
import scala.util.Try

case class Product(price: Double)
case class Item(product: Product, quantity: Int)
case class Order(items: List[Item])

object Main extends App {
  def safeCalculateTotal(order: Order): Option[Double] = {
    Try {
      order.items.map(item => item.product.price * item.quantity).sum
    }.toOption
  }

  val order = Order(List(Item(Product(100.0), 2), Item(Product(50.0), 1)))
  
  safeCalculateTotal(order) match {
    case Some(total) => println(s"Scala: Расчет завершен: $total")
    case None        => println("Scala: Ошибка при расчете")
  }
}

File: lab-6-6/src/main.rs
struct Product { price: f64 }
struct Item { product: Product, quantity: u32 }
struct Order { items: Vec<Item> }

fn type_safe_calculation(order: &Order) -> f64 {
    // Безопасное преобразование u32 -> f64 и строгая типизация суммы
    order.items.iter()
        .map(|item| item.product.price * f64::from(item.quantity))
        .sum()
}

fn main() {
    let order = Order {
        items: vec![
            Item { product: Product { price: 100.0 }, quantity: 2 },
            Item { product: Product { price: 50.0 }, quantity: 1 },
        ],
    };
    let total = type_safe_calculation(&order);
    println!("Rust: Итоговая сумма (безопасно): {:.2}", total);
}